### PR TITLE
std.heap: do not excessively call mmap, and munmap in direct allocator

### PR DESCRIPTION
**Pull request by shawnl**

### CODE REVIEW

**Comment by squeek502** [comment]
This needs to be updated after https://github.com/ziglang/zig/pull/2311.

File: std/heap.zig [hunk 1]
```diff
@@ -144,7 +146,9 @@ pub const DirectAllocator = struct {
                 } else {
                     @memcpy(result.ptr, old_mem.ptr, new_size);
                 }
-                _ = os.posix.munmap(@ptrToInt(old_mem.ptr), old_mem.len);
+                if (old_mem.len > 0) {
```
Comment (daurnimator) [inline std/heap.zig]: This condition could be wrapped around the copying too.

**Comment by squeek502** [comment]
Sorry, introduced conflicts again with https://github.com/ziglang/zig/pull/2357

**Comment by andrewrk** [comment]
:+1:

**Pull request merged**